### PR TITLE
Add transformations for polyhedral fans and complexes

### PR DESCRIPTION
**Pull request by YueRen**
This draft pull request comes from the Leipzig Workshop aims to add the following new operations:

- `Scalar` * `PolyhedralFan` and `PolyhedralFan` * `Scalar` 
- `Scalar` * `PolyhedralComplex` and `PolyhedralFan` * `Scalar`
- -`PolyhedralFan`
- -`PolyhedralComplex`
- `Vector` + `PolyhedralFan` and `PolyhedralFan` + `Vector`
- `Vector` + `PolyhedralComplex` and `PolyhedralComplex` + `Vector`

Questions:

1. Do we want these functions?  My collaborators and I need it for various reasons:
- negation is useful to construct the negated Bergman fan of a matroid (arises for example [here](https://arxiv.org/abs/2212.08173))
- translation is useful to see how the initial ideals of the tropical intersection points vary (relevant for example for [this](https://arxiv.org/abs/2408.15719))

(If yes, I will add the relevant operators for `Cone`, the operators for `Polyhedron` already exist)

2. Is the implementation okay (in particular, I wasn't 100% sure where to put the functions)?

### CODE REVIEW

File: src/PolyhedralGeometry/PolyhedralComplex/standard_constructions.jl [hunk 1]
```diff
@@ -75,3 +75,68 @@ function k_skeleton(PC::PolyhedralComplex{T}, k::Int) where {T<:scalar_types}
...
+
+    # if scalar is non-zero, multiple all vertices and rays by said scalar
+    SigmaVertsAndRays = vertices_and_rays(Sigma)
+    SigmaRayIndices = findall(vr -> vr isa RayVector, SigmaVertsAndRays)
+    SigmaLineality = lineality_space(Sigma)
+    SigmaIncidence = maximal_polyhedra(IncidenceMatrix,Sigma)
+    return polyhedral_complex(SigmaIncidence, multiply_by_nonzero_scalar.(SigmaVertsAndRays,c), SigmaRayIndices, SigmaLineality)
+end
+*(c::Union{ZZRingElem,Rational,Int}, Sigma::PolyhedralComplex) = QQ(c)*Sigma
+*(Sigma::PolyhedralComplex, c::Union{QQFieldElem,ZZRingElem,Rational,Int}) = c*Sigma
```
Comment (lgoettgens) [inline src/PolyhedralGeometry/PolyhedralComplex/standard_constructions.jl]:
```suggestion
*(c::RationalUnion, Sigma::PolyhedralComplex) = QQ(c)*Sigma
*(Sigma::PolyhedralComplex, c::RationalUnion) = c*Sigma
```
this union already has a name

File: src/PolyhedralGeometry/PolyhedralComplex/standard_constructions.jl [hunk 2]
```diff
@@ -75,3 +75,68 @@ function k_skeleton(PC::PolyhedralComplex{T}, k::Int) where {T<:scalar_types}
...
+    return polyhedral_complex(SigmaIncidence, translate_by_vector.(SigmaVertsAndRays,Ref(v)), SigmaRayIndices, SigmaLineality)
+end
++(v::Vector{ZZRingElem}, Sigma::PolyhedralComplex) = QQ.(v)+Sigma
++(v::Vector{Rational}, Sigma::PolyhedralComplex) = QQ.(v)+Sigma
++(v::Vector{Int}, Sigma::PolyhedralComplex) = QQ.(v)+Sigma
+
++(Sigma::PolyhedralComplex, v::Vector{QQFieldElem}) = v+Sigma
++(Sigma::PolyhedralComplex, v::Vector{ZZRingElem}) = QQ.(v)+Sigma
++(Sigma::PolyhedralComplex, v::Vector{Rational}) = QQ.(v)+Sigma
++(Sigma::PolyhedralComplex, v::Vector{Int}) = QQ.(v)+Sigma
```
Comment (lgoettgens) [inline src/PolyhedralGeometry/PolyhedralComplex/standard_constructions.jl]:
```suggestion
+(Sigma::PolyhedralComplex, v::Vector{<:RationalUnion}) = QQ.(v)+Sigma
```

File: src/PolyhedralGeometry/PolyhedralFan/standard_constructions.jl [hunk 3]
```diff
@@ -406,3 +406,40 @@ function arrangement_polynomial(
...
+  end
+
+  # if scalar is non-zero, multiple all vertices and rays by said scalar
+  SigmaRays = first(rays_modulo_lineality(Sigma))
+  SigmaLineality = lineality_space(Sigma)
+  SigmaIncidence = maximal_cones(IncidenceMatrix,Sigma)
+  return polyhedral_fan(SigmaIncidence, multiply_by_nonzero_scalar.(SigmaRays,c), SigmaLineality)
+end
+*(c::Union{ZZRingElem,Rational,Int}, Sigma::PolyhedralFan) = QQ(c)*Sigma
+*(Sigma::PolyhedralFan,c::Union{QQFieldElem,ZZRingElem,Rational,Int}) = c*Sigma
```
Comment (lgoettgens) [inline src/PolyhedralGeometry/PolyhedralFan/standard_constructions.jl]:
```suggestion
*(c::RationalUnion, Sigma::PolyhedralFan) = QQ(c)*Sigma
*(Sigma::PolyhedralFan,c::RationalUnion) = c*Sigma
```

**Comment by benlorenz** [comment]
I think it does make sense to have such operations but they should not be limited to rationals and compatible types but probably use `scalar_types_extended` and then try to coerce the values with `coefficient_field(P)(x)` (for rational fans / complexes this would give QQ but it would also work for number fields and other types).

You can try with `normal_fan(dodecahedron())` as a number field example and `normal_fan(n_gon(7))` as a QQBar example.

_Edit:_ And the first argument for the new object should always be the coefficient field of the old object, e.g. `polyhedral_fan(coefficient_field(f), ...)`.

**Comment by YueRen** [comment]
@benlorenz I tried changing the code as you suggested, but I cannot get it to work.  Can you check why the the following addition is undefined?  

```
julia> Sigma = normal_fan(cube(2))
Polyhedral fan in ambient dimension 2

julia> Sigma + [1,1]
ERROR: MethodError: no method matching +(::PolyhedralFan{QQFieldElem}, ::Vector{Int64})
The function `+` exists, but no method is defined for this combination of argument types.

Closest candidates are:
  +(::Any, ::Any, ::Any, ::Any...)
   @ Base operators.jl:596
  +(::Oscar.StraightLinePrograms.Lazy, ::Any)
   @ Oscar ~/.julia/dev/Oscar/src/StraightLinePrograms/lazy.jl:104
  +(::Any, ::Oscar.StraightLinePrograms.Lazy)
   @ Oscar ~/.julia/dev/Oscar/src/StraightLinePrograms/lazy.jl:105
  ...

Stacktrace:
 [1] top-level scope
   @ REPL[4]:1

```

I was hoping for it to call the following added line in `src/PolyhedralGeometry/PolyhedralComplex/standard_constructions.jl`:
```
+(Sigma::PolyhedralFan, v::Vector{<:scalar_types_extended}) = polyhedral_complex(Sigma)+v
```

**Comment by lgoettgens** [comment]
```julia
julia> Oscar.scalar_types_extended
Union{Float64, FieldElem, ZZRingElem}
```
it seems that `Int` is not part of the type union. Let's hope that @benlorenz has an easy solution for this

**Comment by YueRen** [comment]
In that case, the simplest solution would probably be added all the missing types to `scalar_types_extended`, provided that is what the type was designed for in the first place.

**Comment by benlorenz** [comment]
I had a look at other uses of that union and it should be fine to change it to `Union{scalar_types,IntegerUnion}` to allow `ZZ` and `Integer`.

Edit: it looks like this might need some extra changes for the ray- and pointvectors ...

**Comment by benlorenz** [comment]
from https://github.com/oscar-system/Oscar.jl/pull/4445#issuecomment-2612080011
> @benlorenz I've changed `scalar_types_extended`:
> 
> ```
> -const scalar_types_extended = Union{scalar_types,ZZRingElem}
> +const scalar_types_extended = Union{scalar_types,IntegerUnion}
> ```
> 
> However, it creates a bunch of errors in the type tests that I don't know how to fix: [oscar-system/Oscar.jl/actions/runs/12946690999/job/36111735653?pr=4444#step:8:746](https://github.com/oscar-system/Oscar.jl/actions/runs/12946690999/job/36111735653?pr=4444#step:8:746)
> 
> Can you please look into it when you have the time (most likely not during the currently ongoing polymake developer meeting)? Many thanks in advance!


from https://github.com/oscar-system/Oscar.jl/pull/4445#issuecomment-2623188863
> I am confused, as this PR does not touch `scalar_types_extended` ?!?
> 
> Anyway, the error in the linked log suggest that more things need to be changed if you adjust `scalar_types_extended`. Presumably if an `Int` is now used in places, them at some point some code may call `parent` on it, and this yields `Integers{Int64}()` -- a special placeholder parent for Julia integers that AA provides. And apparently this is then passed as first argument to `point_vector` which doesn't know about it.

Thanks, I am aware of the issue and will take a look.

**Comment by benlorenz** [comment]
The broadcast errors should be fixed now (and I fixed a small issue in the fan negation).

I think most of these constructions would benefit from adding `non_redundant=true` ? (Since they are using non-redundant data from the original objects)

**Comment by fingolfin** [comment]
What the status of this PR?

**Comment by YueRen** [comment]
Sorry for sleeping on this PR, I will look into it and see whether it is mergeable.

**Comment by YueRen** [comment]
@benlorenz @lkastner Can one of you look over the changes and approve the PR (or suggest more changes)?

For reminder, currently OSCAR allows scaling and translating polyhedra (implemented via `Polymake.polytope.translate` and `Polymake.polytope.scale`).  This code tries to implement the same thing for polyhedral complexes.

For that, there added internal functions `scale` and `translate` that each operate on `PointVector` and `RayVector`.  They treat `PointVector` as a normal point in space while `RayVector` is treated as a direction, i.e.: 
* `Oscar.translate(point_vector([1,1]),[2,0]) == point_vector([1,1])`, but 
* `Oscar.translate(ray_vector([1,1]),[2,0]) == ray_vector([1,1])` 
* `Oscar.scale(point_vector([1,1]),-2) == point_vector([-2,-2])`, but 
* `Oscar.scale(ray_vector([1,1]),-2) == ray_vector([-1,-1])` 

In my opinion, it could make sense to overload the `*` and `+` operator with them, but I didn't want to propose such drastic changes.

File: src/PolyhedralGeometry/PolyhedralComplex/standard_constructions.jl [hunk 4]
```diff
@@ -74,3 +74,99 @@ function k_skeleton(PC::PolyhedralComplex{T}, k::Int) where {T<:scalar_types}
...
+    translate.(SigmaVertsAndRays, Ref(v)),
+    SigmaRayIndices,
+    SigmaLineality,
+  )
+end
++(Sigma::PolyhedralComplex, v::Vector{<:scalar_types_extended}) = v + Sigma
+
+# Vector addition for polyhedral fans
++(Sigma::PolyhedralFan, v::Vector{<:scalar_types_extended}) = polyhedral_complex(Sigma) + v
++(v::Vector{<:scalar_types_extended}, Sigma::PolyhedralFan) = v + polyhedral_complex(Sigma)
```
Comment (lkastner) [inline src/PolyhedralGeometry/PolyhedralComplex/standard_constructions.jl]: At some point there was a huge discussion that we don't want to have Minkowski sums of cones and polyhedra. Aren't these two the same? @benlorenz

File: src/PolyhedralGeometry/PolyhedralComplex/standard_constructions.jl [hunk 5]
```diff
@@ -74,3 +74,99 @@ function k_skeleton(PC::PolyhedralComplex{T}, k::Int) where {T<:scalar_types}
...
+  SigmaVertsAndRays = vertices_and_rays(Sigma)
+  SigmaRayIndices = findall(vr -> vr isa RayVector, SigmaVertsAndRays)
+  SigmaLineality = lineality_space(Sigma)
+  SigmaIncidence = maximal_polyhedra(IncidenceMatrix, Sigma)
+  return polyhedral_complex(
+    coefficient_field(Sigma),
+    SigmaIncidence,
+    scale.(SigmaVertsAndRays,c),
+    SigmaRayIndices,
+    SigmaLineality,
```
Comment (benlorenz) [inline src/PolyhedralGeometry/PolyhedralComplex/standard_constructions.jl]:
```suggestion
    SigmaLineality;
    non_redundant=true
```

File: src/PolyhedralGeometry/PolyhedralComplex/standard_constructions.jl [hunk 6]
```diff
@@ -74,3 +74,99 @@ function k_skeleton(PC::PolyhedralComplex{T}, k::Int) where {T<:scalar_types}
...
+## Negation
+###############################################################################
+
+function -(Sigma::PolyhedralComplex)
+  SigmaVertsAndRays = vertices_and_rays(Sigma)
+  SigmaRayIndices = findall(vr -> vr isa RayVector, SigmaVertsAndRays)
+  SigmaLineality = lineality_space(Sigma)
+  SigmaIncidence = maximal_polyhedra(IncidenceMatrix, Sigma)
+  return polyhedral_complex(
+    SigmaIncidence, -SigmaVertsAndRays, SigmaRayIndices, SigmaLineality
```
Comment (benlorenz) [inline src/PolyhedralGeometry/PolyhedralComplex/standard_constructions.jl]:
```suggestion
    coefficient_field(Sigma), SigmaIncidence, -SigmaVertsAndRays, SigmaRayIndices, SigmaLineality; non_redundant=true
```

File: src/PolyhedralGeometry/PolyhedralFan/standard_constructions.jl [hunk 7]
```diff
@@ -405,3 +405,37 @@ function arrangement_polynomial(
   F = parent(first(A))
   return arrangement_polynomial(ring, matrix(F, A); hyperplanes)
 end
+
+###############################################################################
+## Scalar multiplication
+###############################################################################
+
```
Comment (benlorenz) [inline src/PolyhedralGeometry/PolyhedralFan/standard_constructions.jl:412]:
```suggestion

_empty_fan(cf, dim) = polyhedral_fan(cf, incidence_matrix(0, 0), zero_matrix(cf, 0, dim); non_redundant=true)
_origin_fan(cf, dim) = polyhedral_fan(cf, incidence_matrix(1, 0), zero_matrix(cf, 0, dim); non_redundant=true)

```

File: src/PolyhedralGeometry/PolyhedralComplex/standard_constructions.jl [hunk 8]
```diff
@@ -74,3 +74,99 @@ function k_skeleton(PC::PolyhedralComplex{T}, k::Int) where {T<:scalar_types}
...
+)
+  if is_positive(c)
+    return u
+  end
+  if is_negative(c)
+    return -u
+  end
+  return 0*u
+end
+
```
Comment (benlorenz) [inline src/PolyhedralGeometry/PolyhedralComplex/standard_constructions.jl:95]:
```suggestion

_empty_complex(cf, dim) = polyhedral_complex(cf, incidence_matrix(0, 0), zero_matrix(cf, 0, dim); non_redundant=true)
_origin_complex(cf, dim) = polyhedral_complex(cf, incidence_matrix(1, 1, [[1]]), zero_matrix(cf, 1, dim); non_redundant=true)

```

File: src/PolyhedralGeometry/PolyhedralComplex/standard_constructions.jl [hunk 9]
```diff
@@ -74,3 +74,99 @@ function k_skeleton(PC::PolyhedralComplex{T}, k::Int) where {T<:scalar_types}
...
+function *(c::scalar_types_extended, Sigma::PolyhedralComplex)
+  # if scalar is zero, return polyhedral complex consisting only of the origin
+  if iszero(c)
+    return polyhedral_complex(convex_hull(zero_matrix(QQ, 1, ambient_dim(Sigma))))
+  end
```
Comment (benlorenz) [inline src/PolyhedralGeometry/PolyhedralComplex/standard_constructions.jl]:
```suggestion
  n_maximal_polyhedra(Sigma) == 0 && return _empty_complex(coefficient_field(Sigma), ambient_dim(Sigma))
  iszero(c) && return _origin_complex(coefficient_field(Sigma), ambient_dim(Sigma))
```

File: src/PolyhedralGeometry/PolyhedralComplex/standard_constructions.jl [hunk 10]
```diff
@@ -74,3 +74,99 @@ function k_skeleton(PC::PolyhedralComplex{T}, k::Int) where {T<:scalar_types}
...
+  SigmaVertsAndRays = vertices_and_rays(Sigma)
+  SigmaRayIndices = findall(vr -> vr isa RayVector, SigmaVertsAndRays)
+  SigmaLineality = lineality_space(Sigma)
+  SigmaIncidence = maximal_polyhedra(IncidenceMatrix, Sigma)
+  return polyhedral_complex(
+    coefficient_field(Sigma),
+    SigmaIncidence,
+    translate.(SigmaVertsAndRays, Ref(v)),
+    SigmaRayIndices,
+    SigmaLineality,
```
Comment (benlorenz) [inline src/PolyhedralGeometry/PolyhedralComplex/standard_constructions.jl]:
```suggestion
    SigmaLineality;
    non_redundant=true
```

File: src/PolyhedralGeometry/PolyhedralComplex/standard_constructions.jl [hunk 11]
```diff
@@ -74,3 +74,99 @@ function k_skeleton(PC::PolyhedralComplex{T}, k::Int) where {T<:scalar_types}
...
+  )
+end
+*(Sigma::PolyhedralComplex, c::scalar_types_extended) = c * Sigma
+
+###############################################################################
+## Negation
+###############################################################################
+
+function -(Sigma::PolyhedralComplex)
```
Comment (benlorenz) [inline src/PolyhedralGeometry/PolyhedralComplex/standard_constructions.jl:128]:
```suggestion
function -(Sigma::PolyhedralComplex)
  n_maximal_polyhedra(Sigma) == 0 && return _empty_complex(coefficient_field(Sigma), ambient_dim(Sigma))
```

File: src/PolyhedralGeometry/PolyhedralComplex/standard_constructions.jl [hunk 12]
```diff
@@ -74,3 +74,99 @@ function k_skeleton(PC::PolyhedralComplex{T}, k::Int) where {T<:scalar_types}
...
+)
+  return u + v
+end
+function translate(
+  u::RayVector{<:scalar_types_extended}, ::AbstractVector{<:scalar_types_extended}
+)
+  return u
+end
+function +(v::Vector{<:scalar_types_extended}, Sigma::PolyhedralComplex)
+  @req length(v) == ambient_dim(Sigma) "ambient dimension mismatch"
```
Comment (benlorenz) [inline src/PolyhedralGeometry/PolyhedralComplex/standard_constructions.jl:159]:
```suggestion
  @req length(v) == ambient_dim(Sigma) "ambient dimension mismatch"
  n_maximal_polyhedra(Sigma) == 0 && return _empty_complex(coefficient_field(Sigma), ambient_dim(Sigma))
```

File: src/PolyhedralGeometry/PolyhedralFan/standard_constructions.jl [hunk 13]
```diff
@@ -405,3 +405,37 @@ function arrangement_polynomial(
...
+function *(c::scalar_types_extended, Sigma::PolyhedralFan)
+  # if scalar is zero, return polyhedral fan consisting only of the origin
+  if iszero(c)
+    return polyhedral_fan(
+      positive_hull(zero_matrix(coefficient_field(Sigma), 1, ambient_dim(Sigma)))
+    )
+  end
```
Comment (benlorenz) [inline src/PolyhedralGeometry/PolyhedralFan/standard_constructions.jl]:
```suggestion
  n_maximal_cones(Sigma) == 0 && return _empty_fan(coefficient_field(Sigma), ambient_dim(Sigma))
  iszero(c) && return _origin_fan(coefficient_field(Sigma), ambient_dim(Sigma))
```

File: src/PolyhedralGeometry/PolyhedralFan/standard_constructions.jl [hunk 14]
```diff
@@ -405,3 +405,37 @@ function arrangement_polynomial(
...
+    coefficient_field(Sigma), SigmaIncidence, SigmaRays .* c, SigmaLineality
+  )
+end
+*(Sigma::PolyhedralFan, c::scalar_types_extended) = c * Sigma
+
+###############################################################################
+## Negation
+###############################################################################
+
+function -(Sigma::PolyhedralFan)
```
Comment (benlorenz) [inline src/PolyhedralGeometry/PolyhedralFan/standard_constructions.jl]:
```suggestion
function -(Sigma::PolyhedralFan)
  n_maximal_cones(Sigma) == 0 && return _empty_fan(coefficient_field(Sigma), ambient_dim(Sigma))
```

File: src/PolyhedralGeometry/PolyhedralFan/standard_constructions.jl [hunk 15]
```diff
@@ -405,3 +405,37 @@ function arrangement_polynomial(
...
+      positive_hull(zero_matrix(coefficient_field(Sigma), 1, ambient_dim(Sigma)))
+    )
+  end
+
+  # if scalar is non-zero, multiple all rays by said scalar
+  SigmaRays = first(rays_modulo_lineality(Sigma))
+  SigmaLineality = lineality_space(Sigma)
+  SigmaIncidence = maximal_cones(IncidenceMatrix, Sigma)
+  return polyhedral_fan(
+    coefficient_field(Sigma), SigmaIncidence, SigmaRays .* c, SigmaLineality
```
Comment (benlorenz) [inline src/PolyhedralGeometry/PolyhedralFan/standard_constructions.jl]:
```suggestion
    coefficient_field(Sigma), SigmaIncidence, SigmaRays .* c, SigmaLineality; non_redundant=true
```

File: src/PolyhedralGeometry/PolyhedralFan/standard_constructions.jl [hunk 16]
```diff
@@ -405,3 +405,37 @@ function arrangement_polynomial(
...
+
+###############################################################################
+## Negation
+###############################################################################
+
+function -(Sigma::PolyhedralFan)
+  SigmaRays, SigmaLineality = rays_modulo_lineality(Sigma)
+  SigmaIncidence = maximal_cones(IncidenceMatrix, Sigma)
+  return polyhedral_fan(
+    coefficient_field(Sigma), SigmaIncidence, -SigmaRays, SigmaLineality
```
Comment (benlorenz) [inline src/PolyhedralGeometry/PolyhedralFan/standard_constructions.jl]:
```suggestion
    coefficient_field(Sigma), SigmaIncidence, -SigmaRays, SigmaLineality; non_redundant=true
```

**Review by benlorenz — COMMENTED** [review]
I think most (or even all) of the added constructions would benefit from setting `non_redundant=true`.

I would suggest to add helper functions for the corner cases and use them as needed, see all the suggestions. (If we use `non_redundant` then the `iszero` special case is important to make sure only one vertex/ray is passed, but for the empty complex / fan this special case would be wrong in some cases)

Maybe add some tests for non-rational cases, e.g., with a dodecahedron, corner cases like the empty fan / complex and cases triggering the `iszero` lines.

File: src/PolyhedralGeometry/PolyhedralComplex/standard_constructions.jl [hunk 17]
```diff
@@ -74,3 +74,104 @@ function k_skeleton(PC::PolyhedralComplex{T}, k::Int) where {T<:scalar_types}
...
+end
+function scale(
+  u::RayVector{<:scalar_types_extended}, c::scalar_types_extended
+)
+  if is_positive(c)
+    return u
+  end
+  if is_negative(c)
+    return -u
+  end
```
Comment (fingolfin) [inline src/PolyhedralGeometry/PolyhedralComplex/standard_constructions.jl]:
```suggestion
  is_positive(c) && return u
  is_negative(c) && return -u
```

File: src/PolyhedralGeometry/PolyhedralComplex/standard_constructions.jl [hunk 18]
```diff
@@ -74,3 +74,104 @@ function k_skeleton(PC::PolyhedralComplex{T}, k::Int) where {T<:scalar_types}
...
+  )
+end
+
+###############################################################################
+## Translation
+###############################################################################
+
+# helper functions to ensure that point and ray vectors are translated differently:
+# - point vectors represent points in space and should be translated normally
+# - ray vectors represent directions and they should not change at all
```
Comment (fingolfin) [inline src/PolyhedralGeometry/PolyhedralComplex/standard_constructions.jl:147]: In which context is this operation ("translating" a ray vector) useful? Naively to me it would seem safer to have throw an error, as I probably used it by accident. But I suppose you have some applications in mind where it might be useful?

**Review by fingolfin — APPROVED** [review]
Looks good to me

File: src/PolyhedralGeometry/PolyhedralComplex/standard_constructions.jl (hunk 18 above)
Comment (benlorenz) [reply src/PolyhedralGeometry/PolyhedralComplex/standard_constructions.jl:147]: Translating an unbounded polyhedron is achieved by moving the vertices but keeping the same rays. This way it can be directly applied to the list of vertices+rays without filtering. I think it is a reasonable operation like this.

File: src/PolyhedralGeometry/PolyhedralComplex/standard_constructions.jl [hunk 19]
```diff
@@ -74,3 +74,104 @@ function k_skeleton(PC::PolyhedralComplex{T}, k::Int) where {T<:scalar_types}
...
+  u::PointVector{<:scalar_types_extended}, v::AbstractVector{<:scalar_types_extended}
+)
+  return u + v
+end
+function translate(
+  u::RayVector{<:scalar_types_extended}, ::AbstractVector{<:scalar_types_extended}
+)
+  return u
+end
+function +(v::Vector{<:scalar_types_extended}, Sigma::PolyhedralComplex)
```
Comment (benlorenz) [inline src/PolyhedralGeometry/PolyhedralComplex/standard_constructions.jl]: This should probably allow AbstractVector:
```suggestion
function +(v::AbstractVector{<:scalar_types_extended}, Sigma::PolyhedralComplex)
```

File: src/PolyhedralGeometry/PolyhedralComplex/standard_constructions.jl [hunk 20]
```diff
@@ -74,3 +74,104 @@ function k_skeleton(PC::PolyhedralComplex{T}, k::Int) where {T<:scalar_types}
...
+    SigmaRayIndices,
+    SigmaLineality;
+    non_redundant=true
+  )
+end
++(Sigma::PolyhedralComplex, v::Vector{<:scalar_types_extended}) = v + Sigma
+
+# Vector addition for polyhedral fans
++(Sigma::PolyhedralFan, v::Vector{<:scalar_types_extended}) = polyhedral_complex(Sigma) + v
++(v::Vector{<:scalar_types_extended}, Sigma::PolyhedralFan) = v + polyhedral_complex(Sigma)
```
Comment (benlorenz) [inline src/PolyhedralGeometry/PolyhedralComplex/standard_constructions.jl]:
```suggestion
+(Sigma::PolyhedralComplex, v::AbstractVector{<:scalar_types_extended}) = v + Sigma

# Vector addition for polyhedral fans
+(Sigma::PolyhedralFan, v::AbstractVector{<:scalar_types_extended}) = polyhedral_complex(Sigma) + v
+(v::AbstractVector{<:scalar_types_extended}, Sigma::PolyhedralFan) = v + polyhedral_complex(Sigma)
```

File: src/PolyhedralGeometry/PolyhedralComplex/standard_constructions.jl [hunk 21]
```diff
@@ -74,3 +74,99 @@ function k_skeleton(PC::PolyhedralComplex{T}, k::Int) where {T<:scalar_types}
...
+  u::RayVector{<:scalar_types_extended}, c::scalar_types_extended
+)
+  if is_positive(c)
+    return u
+  end
+  if is_negative(c)
+    return -u
+  end
+  return 0*u
+end
```
Comment (lkastner) [inline src/PolyhedralGeometry/PolyhedralComplex/standard_constructions.jl:94]: Is `sign` available for all `scalar_types_extended`? If so you could shorten it a bit:
```suggestion
function scale(
  u::RayVector{<:scalar_types_extended}, c::scalar_types_extended
)
  if is_positive(c)
    return u
  else
  	return sign(c)*u
  end
end
```
It could also be shortened more, but maybe it gets slower then?

File: src/PolyhedralGeometry/PolyhedralComplex/standard_constructions.jl [hunk 22]
```diff
@@ -74,3 +74,99 @@ function k_skeleton(PC::PolyhedralComplex{T}, k::Int) where {T<:scalar_types}
...
+###############################################################################
+
+# helper functions to ensure that point and ray vectors scale differently:
+# - point vectors represent points in space and should be scaled normally
+# - ray vectors represent directions and they should only be scaled by 0 or +/-1
+function scale(
+  u::PointVector{<:scalar_types_extended}, c::scalar_types_extended
+)
+  return u * c
+end
```
Comment (lkastner) [inline src/PolyhedralGeometry/PolyhedralComplex/standard_constructions.jl]:
```suggestion
scale(
  u::PointVector{<:scalar_types_extended}, c::scalar_types_extended
) = u * c
```

File: src/PolyhedralGeometry/PolyhedralFan/standard_constructions.jl [hunk 23]
```diff
@@ -405,3 +405,37 @@ function arrangement_polynomial(
...
+    )
+  end
+
+  # if scalar is non-zero, multiple all rays by said scalar
+  SigmaRays = first(rays_modulo_lineality(Sigma))
+  SigmaLineality = lineality_space(Sigma)
+  SigmaIncidence = maximal_cones(IncidenceMatrix, Sigma)
+  return polyhedral_fan(
+    coefficient_field(Sigma), SigmaIncidence, SigmaRays .* c, SigmaLineality
+  )
```
Comment (lkastner) [inline src/PolyhedralGeometry/PolyhedralFan/standard_constructions.jl]: This I do not understand. I would have expected this to behave similarly to `scale`, i.e. just return the object itself if `c>0` and otherwise negate the rays. Something like:
```suggestion
  if is_positive(c)
    return Sigma
  else
    return -Sigma
  end
```

File: src/PolyhedralGeometry/PolyhedralComplex/standard_constructions.jl (hunk 21 above)
Comment (lkastner) [reply src/PolyhedralGeometry/PolyhedralComplex/standard_constructions.jl:94]: Sorry, a train ride interrupted my review, so this is outdated.

File: src/PolyhedralGeometry/PolyhedralComplex/standard_constructions.jl (hunk 4 above)
Comment (YueRen) [reply src/PolyhedralGeometry/PolyhedralComplex/standard_constructions.jl]: True, I vaguely remember the decision that OSCAR should never consider polyhedral fans as a special case of polyhedral complexes.  If that is true, then the addition operator for fans should be removed (as it doesn't make sense for polyhedral fans).

Comment (YueRen) [reply src/PolyhedralGeometry/PolyhedralComplex/standard_constructions.jl]: Once I find the time, hopefully today, I'll check the old discussion and adjust this PR accordingly.

Comment (benlorenz) [reply src/PolyhedralGeometry/PolyhedralComplex/standard_constructions.jl]: The other PR was: https://github.com/oscar-system/Oscar.jl/pull/4828
I do agree that it would be better to not have these two lines, it should be easy enough to explicitly cast them first.

Comment (YueRen) [reply src/PolyhedralGeometry/PolyhedralComplex/standard_constructions.jl]: I've remove the functions.

File: src/PolyhedralGeometry/PolyhedralFan/standard_constructions.jl (hunk 23 above)
Comment (YueRen) [reply src/PolyhedralGeometry/PolyhedralFan/standard_constructions.jl]: You are right, it is simplified now.

**Comment by fingolfin** [comment]
The formatter complains now, but that could be addressed by running `julia etc/format_code.jl` (and even someone else can run it for you, @YueRen).

The main question is whether @lkastner and @benlorenz are happy with the the general content now, the formatting is an easy afterthought.

**Comment by YueRen** [comment]
> The formatter complains now, but that could be addressed by running `julia etc/format_code.jl` (and even someone else can run it for you, @YueRen).
> 
> The main question is whether @lkastner and @benlorenz are happy with the the general content now, the formatting is an easy afterthought.

I've run the formatting script and pushed the changes.

**Review by lkastner — APPROVED** [review]
Thank you. We should probably expand `transform` on the polymake side though.

File: test/PolyhedralGeometry/polyhedral_complex.jl [hunk 24]
```diff
@@ -151,4 +151,28 @@
...
+    @test ambient_dim(PCscaled) == ambient_dim(PCshifted)
+    @test lineality_dim(PCscaled) == lineality_dim(PCshifted)
+    @test issetequal(rays(PCscaled), rays(PCshifted))
+    @test n_maximal_polyhedra(PCscaled) == n_maximal_polyhedra(PCshifted)
+
+    PCnegated = -PCshifted
+    @test dim(PCnegated) == dim(PCshifted)
+    @test ambient_dim(PCnegated) == ambient_dim(PCshifted)
+    @test lineality_dim(PCnegated) == lineality_dim(PCshifted)
+    @test issetequal(rays(PCnegated), rays(PCshifted))
```
Comment (benlorenz) [inline test/PolyhedralGeometry/polyhedral_complex.jl]:
```suggestion
    @test issetequal(rays(PCnegated), -1 .* rays(PCshifted))
```
Otherwise this is wrong as soon as the input polytope is not as symmetric as the cube.

File: test/PolyhedralGeometry/polyhedral_complex.jl [hunk 25]
```diff
@@ -151,4 +151,28 @@
       @test n_maximal_polyhedra(vrep) == n_maximal_polyhedra(hrep)
     end
   end
+
+  @testset "Operators" begin
+    PC = polyhedral_complex(normal_fan(cube(2)))
+    PCshifted = PC + [1, 1]
```
Comment (benlorenz) [inline test/PolyhedralGeometry/polyhedral_complex.jl]: Test some more examples:
```suggestion
  @testset "Transformations" for P in (cube(2), dodecahedron(), n_gon(5))
    PC = polyhedral_complex(normal_fan(P))
    PCshifted = PC + fill(1, dim(P))
```
This uncovered the rays issue below, and these three objects cover the most important types that we support (QQ, number fields, QQBar).

File: test/PolyhedralGeometry/polyhedral_fan.jl [hunk 26]
```diff
@@ -222,6 +222,17 @@ end
   @test number_of_maximal_cones(sff1) == 3
 end
 
+@testset "Operations" begin
+  f = normal_fan(cube(2))
```
Comment (benlorenz) [inline test/PolyhedralGeometry/polyhedral_fan.jl]:
```suggestion
@testset "Transformations" for P in (cube(2), dodecahedron(), n_gon(5))
  f = normal_fan(P)
```

**Review by benlorenz — CHANGES_REQUESTED** [review]

**Review by benlorenz — APPROVED** [review]
